### ISSUE TITLE
Architecture & training design issues for future work

### ISSUE THREAD

**Issue by Jairaj1234-dancer** [sub-surface/hexgo#8]
# Architecture & Training Design Issues

Findings from a deep multi-agent code review that are beyond the scope of the 6 PRs submitted (#2–#7). These are structural design issues that require deliberate architectural decisions rather than targeted fixes.

---

## 1. Per-move policy head is O(M) per position — fundamental scaling bottleneck

**Files:** `net.py:432-441`, `inference.py:244-247`

Each legal move requires its own row in the batch tensor — the board is duplicated N times (one per legal move). With ~50-100 legal moves in ZOI, a single MCTS evaluation inflates the batch by 50-100x. A batch of 8 requests with 30 moves each creates 240 trunk forward passes instead of 8.

This makes trunk computation cost **O(M) per position** instead of O(1). AlphaZero and KataGo both use a **spatial policy head** that outputs logits for all board positions in one pass — the trunk runs once, then a 1×1 conv produces a `[B, 1, S, S]` policy map, and legal moves are read off by coordinate lookup.

**Impact:** This is the single largest performance bottleneck. At higher sim counts or larger networks, inference cost will be dominated by redundant trunk computation. The current design makes it impractical to scale beyond ~200 sims/move.

**Recommendation:** Replace the per-move-plane policy head with a spatial policy head. The move plane concatenation was designed to avoid a fixed output size on the infinite grid, but the 18×18 window already bounds the output — a `[18, 18]` spatial policy map with coordinate lookup achieves the same flexibility at 50-100x lower cost.

---

## 2. TD-lambda targets are not actually TD-lambda

**File:** `train.py:318-328`

The code comments say "TD-lambda" but the formula `z_t = gamma^(T-1-t) * z_final` is **exponential discounting of the final outcome**, not TD-lambda (which involves bootstrapped multi-step returns mixing TD(0) through TD(T)).

This formula makes early positions' targets exponentially close to zero:
- 60-move game, position 0: `0.99^59 = 0.55` of final outcome (moderate)
- 150-move game, position 0: `0.99^149 = 0.22` of final outcome (severe)
- 300-move game, position 0: `0.99^299 = 0.05` (near-zero)

**Risk:** The value head learns to predict ~0 for all opening positions regardless of game-theoretic value. This is a **value target collapse** for long games. The network will systematically undervalue early-game decisions.

**Recommendation:** Either:
1. Use true TD(λ) with bootstrapped returns: `z_t = (1-λ) Σ_{n=1}^{T-t} λ^{n-1} G_t^{(n)}` where `G_t^{(n)}` mixes n-step returns with value estimates
2. Or simply use uniform `z = ±1` (original AlphaZero approach) — simpler and proven to work

---

## 3. Auxiliary labels leak spatial future information

**Files:** `train.py:349`, `net.py:470-481`

Ownership and threat labels are generated from the **final game state** but applied to **every training position** in the episode. The code acknowledges this (line 479-481), analogizing to AlphaZero's value target.

However, value targets are scalars (win/loss), while ownership is a **spatial map** — predicting which specific cells will be occupied at move 5 when the label shows pieces placed at move 40+ is a much stronger future-information signal. The threat label similarly shows the exact winning line, which may not even exist as a threat at early positions.

**Risk:** The trunk may overfit to predicting game trajectories (which cells will be played) rather than learning positional evaluation (which positions are strategically good). This could make the trunk representations brittle and trajectory-dependent rather than generalizable.

**Recommendation:** Consider position-specific aux labels:
- **Ownership at current state** rather than final state (which cells are currently occupied)
- **Threat relative to current position** (which cells are part of existing chains of length ≥ 4)
- Or reduce `AUX_LOSS_OWN` and `AUX_LOSS_THREAT` further (currently 0.1 each) to minimize the future-leakage signal

---

## 4. Board centering breaks on split-front play

**File:** `net.py:179-188`

When `N_RECENT=20` moves are split between two distant fronts (e.g., moves 1-10 in one area, moves 11-20 in another), the centroid falls **between** them. The 18×18 window then shows empty space in the center with partial views of both fronts — the network loses critical information from both active areas.

This is a natural consequence of Connect-6 on an infinite grid: players can open multiple threats in distant regions to force the opponent to respond in one while pressing the other.

**Recommendation:** Options:
1. **Attention-based centering** — center on the front with the highest threat density rather than the centroid
2. **Multi-window encoding** — encode two 18×18 windows (one per active front) and fuse features
3. **Larger window** — increase BOARD_SIZE to 24 or 32 (increases compute quadratically but handles split fronts up to ~12 hex distance apart)
4. **Dynamic window** — detect split fronts and encode the more tactically urgent one

---

## 5. Persistent inference cache serves stale evaluations across generations

**File:** `inference.py:34-48`

The persistent cache (`_persistent_cache`) retains (value, policy) pairs for `CACHE_MAX_AGE=5` generations. After the network updates, these are **stale** — they reflect the old network's beliefs, not the current one.

Serving stale policy priors during MCTS corrupts the search: the tree explores moves the old network liked rather than what the current network recommends. Stale values are even worse — the search accumulates value estimates from a network that may have had fundamentally different positional understanding.

The per-generation server cache is also populated **from** the persistent cache at line 176-178 without staleness checking, so stale entries propagate into the fast path.

**Impact:** Frequently-visited early-game positions are most likely to be cached and therefore most affected. These are precisely the positions where the network's understanding changes most dramatically between generations.

**Recommendation:** Either:
1. Reduce `CACHE_MAX_AGE` to 1 (only cache within the current generation)
2. Or invalidate the persistent cache on weight sync (clear it when `load_state_dict` is called)
3. Or tag entries with a weight version hash and reject mismatches

---

## 6. Effective receptive field may be too small for 6-in-a-row detection

**File:** `net.py:34-35, 269-280`

With 4 ResBlocks (8 conv layers) + 1 stem conv = 9 conv layers, the **theoretical** receptive field is 19×19 — barely covering the 18×18 window. But the **effective** receptive field (ERF) for small networks is typically 1/3 to 1/2 of theoretical, giving an ERF of ~6-10 cells.

The `GlobalPoolBranch` mitigates this by broadcasting aggregate statistics (mean/max), but it provides only global summary info, not positional awareness. A 6-in-a-row threat spanning 12+ cells diagonally likely exceeds the ERF — the network may not be able to "see" both ends of a near-winning chain simultaneously.

**Recommendation:** Options:
1. Add dilated convolutions (e.g., dilation=2 in one ResBlock) to expand ERF without adding layers
2. Increase to 6-8 blocks (more compute but directly solves the problem)
3. Add a lightweight self-attention layer after the trunk (expensive but maximally flexible)

---

## 7. CA Laplacian init sets identical weights across all channels

**File:** `net.py:336-347`

Every `HexConv2d` filter starts as the same Laplacian kernel (center=+1, neighbors=-1/6). The Xavier scale factor at line 346 uses `fan = w[0].numel()` giving `scale ~= 0.059` for 64-channel ResBlock convs, making all initial weights very small and nearly identical.

Symmetry breaking depends entirely on BatchNorm noise and optimizer stochasticity. This could slow early training convergence compared to standard Kaiming init with random diversity, where each filter starts with a different random pattern.

**Recommendation:** Add small random perturbation to the Laplacian template:
```python
w.mul_(scale)
w.add_(torch.randn_like(w) * scale * 0.1)  # break symmetry
```

---

## 8. D6 augmentation is not perfectly equivalent for border positions

**File:** `net.py:67-106`

For non-identity D6 transforms, some source pixels map outside the 18×18 window (line 91 validity check) and their information is **lost** (zeroed in destination). Because the window is square but the D6 group includes 60° rotations, the information loss is **different** for each transform — corner regions are clipped differently under rotation vs reflection.

This means the 12 augmented samples are **not truly equivalent**, adding label noise proportional to the fraction of occupied border cells. For positions with pieces near the window edge (increasingly common in mid-to-late game), this noise can be significant.

**Recommendation:** 
1. Filter augmented samples where >5% of occupied cells fall outside the window
2. Or use a circular mask centered on the window to make clipping rotationally symmetric
3. Or increase window size to reduce border effects (most pieces cluster in the center anyway)

---

## Priority ordering

| # | Issue | Impact | Effort |
|---|-------|--------|--------|
| 1 | Spatial policy head | 50-100x inference speedup | Large (architecture change) |
| 5 | Stale cache invalidation | Corrupted search quality | Small (clear on sync) |
| 2 | TD-lambda → uniform targets | Value head accuracy | Small (remove discounting) |
| 6 | Receptive field | Threat detection range | Medium (add dilated convs) |
| 4 | Split-front centering | Information loss | Medium (needs design) |
| 3 | Aux label leakage | Representation quality | Medium (needs design) |
| 7 | CA init symmetry breaking | Early training speed | Small (add noise) |
| 8 | D6 border clipping | Label noise | Small (filter or mask) |

*Filed from comprehensive code review. Related PRs: #2, #3, #4, #5, #6, #7.*
